### ISSUE TITLE
Deploy precompiled binaries to Lambda to prevent cold starts

### ISSUE THREAD

**Issue by r007** [ocius/api#7]
@thomasdane I come up with an idea of how we can improve performance of our backend. Some of the reasons cold starts are slow, such as copying your code from S3, apply to all of the supported runtimes, not just .NET Core. The biggest performance hit for .NET functions, however, comes from converting the Common Intermediate Language (CIL) code into machine code via just-in-time (JIT) compilation.

The more code and libraries you use in your function, the slower it will be. Less abstraction and more action leads to faster cold starts.

Let's try to use [LambdaNative](https://github.com/zaccharles/lambda-native) library to pre-compile C# code into binaries. Here's a good article about it: [Making .NET AWS Lambda Functions Start 10x Faster using LambdaNative](https://medium.com/@zaccharles/making-net-aws-lambda-functions-start-10x-faster-using-lambdanative-8e53d6f12c9c)

Also, here's the practical example of how to implement it:
[lambda-native/example](https://github.com/zaccharles/lambda-native/tree/master/example)

What do you think?

**Comment by thomasdane** [comment]
Great find. Will take a look in the new year!

**Comment by thomasdane** [comment]
@r007 I don't know if cold starts are the problem. Even on repeated calls, API Gateway takes ~600ms to respond. Check out the following screenshot: 

It shows that the response took about 600ms. This was on the third or fourth request I made - the container was already running. The problem seems not to be cold starts, but simple response time. The request goes API Gateway --> Lambda --> DynamoDB and back.

**Comment by r007** [comment]
@thomasdane yeah, I've seen that. It's a fairly simple query, normally it should take 100-200ms. But when I query the daily locations, the response time raises dramatically. From 600ms to around 1000ms. So the more complicated it gets, the slower it will be. It might look insignificant at this moment, but later it may get terribly slow. It diminishes all the work that we've done to improve performance of the live page. Because the react app itself utterly depends on the API.

Can you still try my suggestion?

**Comment by thomasdane** [comment]
There is definitely a bottle neck in the AWS response, but I don't thin it is lamda. Look at the response time when getting locations by day: 

![locationsDay](https://user-images.githubusercontent.com/2957318/72590535-d2314580-38b2-11ea-9394-f853d7d37a3b.png)

and drones by day: 

![dronesDay](https://user-images.githubusercontent.com/2957318/72590689-3522dc80-38b3-11ea-98a8-b699a84977ce.png)

The lambda function returns in about 100ms. The bottleneck is thus API gateway, or the browser. 

Let's book some time to chat about it over the phone.

**Comment by r007** [comment]
@thomasdane hard to say anything, since I haven't seen your configuration. It's a black box for me. But let's assume that lambda is working as should. Then what else it could be?

It could be something obvious, let's say physical distance. If you created resources in remote places of the world. The database in California and lambda is in Europe. So-called cross-region function calls.

Or it could be an issue with API Gateway https://github.com/serverless/serverless/issues/139.

Also, it can be because of encryption/decryption of caches (If you have request caching with encryption enabled). Or because of role-based authentication (IAM Roles are slower than resource-based authentication).

We can try to exclude theories one by one.
